x/gov/keeper/participation_ema.go

```
@@ -0,0 +1,110 @@
+package keeper
+
+import (
+	"context"
+	"errors"
+
+	"cosmossdk.io/collections"
+	"cosmossdk.io/math"
+
+	v1 "github.com/cosmos/cosmos-sdk/x/gov/types/v1"
+)
+
+// UpdateParticipationEMA updates the governance participation EMA
+func (k Keeper) UpdateParticipationEMA(ctx context.Context, proposal v1.Proposal, participation math.LegacyDec) {
+	formula := func(old, new math.LegacyDec) math.LegacyDec {
+		return old.Mul(math.LegacyNewDecWithPrec(8, 1)).Add(new.Mul(math.LegacyNewDecWithPrec(2, 1)))
+	}
+
+	kinds := k.ProposalKinds(proposal)
+	if kinds.HasKindConstitutionAmendment() {
+		current, err := k.ConstitutionAmendmentParticipationEMA.Get(ctx)
+		if err != nil {
+			panic(err)
+		}
+
+		if err := k.ConstitutionAmendmentParticipationEMA.Set(ctx, formula(current, participation)); err != nil {
+			panic(err)
+		}
+	}
+	if kinds.HasKindLaw() {
+		current, err := k.LawParticipationEMA.Get(ctx)
+		if err != nil {
+			panic(err)
+		}
+
+		if err := k.LawParticipationEMA.Set(ctx, formula(current, participation)); err != nil {
+			panic(err)
+		}
+	}
+	if kinds.HasKindAny() {
+		current, err := k.ParticipationEMA.Get(ctx)
+		if err != nil {
+			panic(err)
+		}
+
+		if err := k.ParticipationEMA.Set(ctx, formula(current, participation)); err != nil {
+			panic(err)
+		}
+	}
+}
+
+// GetQuorum returns the dynamic quorum for governance proposals calculated
+// based on the participation EMA
+func (k Keeper) GetQuorum(ctx context.Context) math.LegacyDec {
+	params, err := k.Params.Get(ctx)
+	if err != nil {
+		panic(err)
```

shouldn't we propagate the error here since this method and the others below are called inside `EndBlocker` ?

---

Any error in an EndBlock will endup in a chain halt. returning errors is more idomatic tho, so I can update. This is mainly a copy/paste + collection update

---

> Any error in an EndBlock will endup in a chain halt.

Yes that's my point. Thanks for updating!

---

Right, i looked at the current code path in atom one, and it looks more invalid to me:

- https://github.com/atomone-hub/atomone/blob/bbcaae1/x/gov/keeper/participation_ema.go#L78-L86
- https://github.com/atomone-hub/atomone/blob/bbcaae1/x/gov/keeper/params.go#L22-L32

If the key is not found we return an empty struct, meaning the quorum will be at 0, and the proposal may pass and get executed. I believe if the key is not found, we should probably halt.